handler/rfc8628/strategy_hmacsha_test.go

```diff
@@ -0,0 +1,131 @@
+// Copyright © 2023 Ory Corp
+// SPDX-License-Identifier: Apache-2.0
+
+package rfc8628_test
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/patrickmn/go-cache"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/ory/fosite"
+	. "github.com/ory/fosite/handler/rfc8628"
+	"github.com/ory/fosite/token/hmac"
+)
+
+var hmacshaStrategy = DefaultDeviceStrategy{
+	Enigma:           &hmac.HMACStrategy{Config: &fosite.Config{GlobalSecret: []byte("foobarfoobarfoobarfoobarfoobarfoobarfoobarfoobar")}},
+	RateLimiterCache: cache.New(24*time.Minute, 2*24*time.Minute),
+	Config: &fosite.Config{
+		AccessTokenLifespan:            time.Minute * 24,
+		AuthorizeCodeLifespan:          time.Minute * 24,
+		DeviceAndUserCodeLifespan:      time.Minute * 24,
+		DeviceAuthTokenPollingInterval: 400 * time.Millisecond,
+	},
+}
+
+var hmacValidCase = fosite.Request{
+	Client: &fosite.DefaultClient{
+		Secret: []byte("foobarfoobarfoobarfoobar"),
+	},
+	Session: &fosite.DefaultSession{
+		ExpiresAt: map[fosite.TokenType]time.Time{
+			fosite.UserCode:   time.Now().UTC().Add(time.Hour),
+			fosite.DeviceCode: time.Now().UTC().Add(time.Hour),
+		},
+	},
+}
+
+func TestHMACUserCode(t *testing.T) {
+	for k, c := range []struct {
+		r    fosite.Request
+		pass bool
+	}{
+		{
+			r:    hmacValidCase,
+			pass: true,
+		},
+		// TODO: add tests for expired codes
```

Comment:
perhaps create Jira tasks for the two TODOs in this file

Reply:
There is a jira ticket for this, the work will be done after this PR is merged